flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/Splitter.scala

```diff
@@ -0,0 +1,215 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.ml.preprocessing
+
+import org.apache.flink.api.common.typeinfo.{TypeInformation, BasicTypeInfo}
+import org.apache.flink.api.java.Utils
+import org.apache.flink.api.scala. DataSet
+import org.apache.flink.api.scala.utils._
+
+import org.apache.flink.ml.common.{FlinkMLTools, ParameterMap, WithParameters}
+import _root_.scala.reflect.ClassTag
+
+object Splitter {
+
+  case class TrainTestDataSet[T: TypeInformation : ClassTag](training: DataSet[T],
+                                                             testing: DataSet[T])
+
+  case class TrainTestHoldoutDataSet[T: TypeInformation : ClassTag](training: DataSet[T],
+                                                                    testing: DataSet[T],
+                                                                    holdout: DataSet[T])
+  // --------------------------------------------------------------------------------------------
+  //  randomSplit
+  // --------------------------------------------------------------------------------------------
+  /**
+   * Split a DataSet by the probability fraction of each element.
+   *
+   * @param input           DataSet to be split
+   * @param fraction        Probability that each element is chosen, should be [0,1] without
+   *                        replacement, and [0, ∞) with replacement. While fraction is larger
+   *                        than 1, the elements are expected to be selected multi times into
+   *                        sample on average. This fraction refers to the first element in the
+   *                        resulting array.
+   * @param precise         Sampling by default is random and can result in slightly lop-sided
+   *                        sample sets. When precise is true, equal sample set size are forced,
+   *                        however this is somewhat less efficient.
+   * @param seed            Random number generator seed.
+   * @return An array of two datasets
+   */
+
+  def randomSplit[T: TypeInformation : ClassTag]( input: DataSet[T],
+                                                  fraction: Double,
+                                                  precise: Boolean = false,
+                                                  seed: Long = Utils.RNG.nextLong())
+  : Array[DataSet[T]] = {
+    import org.apache.flink.api.scala._
+
+    val indexedInput: DataSet[(Long, T)] = input.zipWithIndex
+
+    val leftSplit: DataSet[(Long, T)] = precise match {
+      case false => indexedInput.sample(false, fraction, seed)
```

Comment:
I think boostrapping would be a cool feature- but would require a different approach than the joins on the leftSplit/rightSplit. 

If you over sample the leftSplit, there's not going to be anything left to put in the right split (the whole points was to keep the training and testing cases seperate).

I'm going to to add a boostrap method that will allow for oversampling in the testing and training cases.  Re: the next comment, I will test is separately.